### DISCUSSION TITLE
Data Fetching with API from the same project

### DISCUSSION THREAD

**Discussion by tofatoff** [vercel/next.js#71887] [Help]
### Summary

I wonder if it's a good practice to make API endpoints to separate data fetching so it's not all relied to the server components to retrieve the data, for instance, from a database. So, the server components are making an API request to the API endpoints made from the same project. I think it's also good for scalability too (like we can use the same API in case we want to make a different web/mobile apps)

**Comment by aniruddhaadak80** [comment]
Hi @tofatoff ,  

Yes, it’s a good practice to create API endpoints for data fetching within the same project. This approach promotes separation of concerns, enhances scalability, and allows for easier reuse across different applications (web/mobile). It also simplifies testing and debugging.  

Thanks!

**Comment by robahtou** [comment]
It depends on the use case.

If another client will make the same data call, a server action or function isn't the right approach.

These are distinct concerns: one is about giving external clients access, while the other focuses on access for your own client. Your client should have the most direct route to the data source, where an RSC (React Server Component) works well. For external clients, however, an RSC isn’t a suitable solution. This scenario is better handled by creating an endpoint in the `/api` folder within your Next.js project. Here, supporting both access types is perfectly acceptable.

In either case, the discussion revolves around "access" (whether through a distinct endpoint or an RSA). And if both methods are pulling identical data from the same source, then shared functions can be implemented for both access points. However, having an RSA make a REST call to the same server where it's already running is unnecessary and redundant.

**Comment by icyJoseph** [comment]
Using the handlers, as the core abstraction to fetch data, is an anti-pattern. The Next.js documents used to be rather harsh on this kind of practice too.

The Next.js discord collected some thoughts on this FAQ:

- https://nextjs-faq.com/fetch-api-in-getserversideprops
- https://nextjs-faq.com/fetch-api-in-rsc

Note that the issue I am pointing at is, using these API endpoints, to fetch data from your own application. Obviously, these endpoints are a valid tool, when you want to give your browser clients, or other parties really, a way to interact with a service.  As you say:

>  in case we want to make a different web/mobile apps

In that case though, your API endpoint, should just implement a piece of core logic. For example:

```ts
// lib/codewars.ts
export const getCodeWarsUser = (user) =>
  db
    .getUser(user)
```

- Then in a server component

```tsx
// pages/index.ts
export default async function Page() {
  const codewars = await getCodeWarsUser('some-user');

  return <div>{/* some stuff based of data */}</div>
}
```

- And also in `api/codewars`, so I can reach it from the client, or any other http client really

```ts
// api/codewars.ts
const codewars = async (req, res) => {
  const data = await getCodeWarsUser('some-user');

  return res.send(data);
};

export default codewars;
```

or in route handlers:

```js

export function GET() {
  const data = await getCodeWarsUser('some-user');

  return Response.json(data)
}
```

However, using the API route, or Route Handler, as the core abstraction, has issues. The biggest issues I see, are that:

- builds need a bit of a catch-22 situation to be properly, and carefully, set-up
- that you need to use absolute URLs.
- an extra http jump, out to the internet from your component, back to your server, which means things will be slower than necessary

**Reply by robahtou** [reply]
The `lib/codewars.ts` needs to updated to be the 'core abstraction' (assuming you own the data source):
```javascript
// if your server can access data directly from the database
export const getCodeWarsUsersById = (user) => {
  // whatever your database library is
  const data = sql`
    SELECT * FROM codewars_users WHERE user = ${user}
  `;
  return data;
};
```

If a client that you don't own (like another teams mobile application or a separate service or whatever), then you can expose that data access via the `api/codewars` via a route handler 

```javascript
export async function GET() {
  const data = await getCodeWarsUsersById('some-user');

  return Response.json(data)
}
```

The core abstraction is the function call that actually 'gets' the data, not the `http` nor the `pages/index.ts`. These are 'access' points to the data.

**Reply by icyJoseph** [reply]
Fair point, this was adapted from an answer I gave several years ago ~ will update the snippets
